parser/prism/Parser.h

```diff
@@ -0,0 +1,104 @@
+#ifndef SORBET_PARSER_PRISM_PARSER_H
+#define SORBET_PARSER_PRISM_PARSER_H
+
+#include <memory>
+#include <string>
+#include <vector>
+
+extern "C" {
+#include "prism.h"
+}
+
+#include "core/LocOffsets.h"
+
+namespace sorbet::parser::Prism {
+
+class Node;
+
+class ParseError {
+public:
+    ParseError(pm_diagnostic_id_t id, const std::string &message, pm_location_t location, pm_error_level_t level)
+        : id(id), message(message), location(location), level(level) {}
+
+    pm_diagnostic_id_t id;
+    std::string message;
+    pm_location_t location;
+    pm_error_level_t level;
+};
+
+// A backing implemenation detail of `Parser`, which stores a Prism parser and its options in a single allocation.
+struct ParserStorage {
+    // The version of Ruby syntax that we're parsing with Prism. This determines what syntax is supported or not.
+    static constexpr std::string_view ParsedRubyVersion = "3.3.0";
+    pm_parser_t parser;
+    pm_options_t options;
+
+    ParserStorage(std::string_view source_code) : parser{}, options{} {
+        pm_options_version_set(&options, ParsedRubyVersion.data(), ParsedRubyVersion.size());
+
+        pm_parser_init(&parser, reinterpret_cast<const uint8_t *>(source_code.data()), source_code.size(), &options);
+    }
+
+    ~ParserStorage() {
+        pm_parser_free(&parser);
+        pm_options_free(&options);
+    }
+
+    ParserStorage(const ParserStorage &) = delete;
+    ParserStorage &operator=(const ParserStorage &) = delete;
+    ParserStorage(ParserStorage &&) = delete;
+    ParserStorage &operator=(ParserStorage &&) = delete;
+};
+
+class Parser final {
+    friend class Node;
+    friend struct NodeDeleter;
+
+    std::shared_ptr<ParserStorage> storage;
+
+public:
+    std::vector<ParseError> parseErrors;
```

Comment:
This doesn't feel like the appropriate place to store this, especially since it's meaningless until `collectErrors` is called.  Perhaps `parse_root` should return both the node and any accumulated errors, and then the translator should take any accumulated parse errors from that, rather than grabbing them off the `Parser`?